Sources/SourceKitLSP/Swift/DiagnosticReportManager.swift

```diff
@@ -0,0 +1,175 @@
+      )
+      // If we don't have build settings,
+      // sourcekitd won't be able to give us accurate semantic diagnostics.
+      // Fall back to providing syntactic diagnostics from the built-in
+      // swift-syntax. That's the best we can do for now.
+      let report = try await requestFallbackReport(with: snapshot)
+      setReport(for: snapshot.id, buildSettings: nil, report: report)
+      return report
+    }
+    guard !buildSettings.isFallback else {
+      logger.log(
+        "Producing syntactic diagnostics from the built-in swift-syntax because we have fallback arguments"
+      )
+      // If we only have fallback build settings,
+      // sourcekitd won't be able to give us accurate semantic diagnostics.
+      // Fall back to providing syntactic diagnostics from the built-in
+      // swift-syntax. That's the best we can do for now.
+      let report = try await requestFallbackReport(with: snapshot)
+      setReport(for: snapshot.id, buildSettings: buildSettings, report: report)
+      return report
+    }
+    let report = try await requestReport(with: snapshot, compilerArgs: buildSettings.compilerArgs)
+    setReport(for: snapshot.id, buildSettings: buildSettings, report: report)
+    return report
+  }
+}
+
+private extension DiagnosticReportManager {
+  func requestReport(with snapshot: DocumentSnapshot, compilerArgs: [String]) async throws
+    -> LanguageServerProtocol.RelatedFullDocumentDiagnosticReport
+  {
+    try Task.checkCancellation()
+
+    let keys = self.keys
+
+    let skreq = SKDRequestDictionary(sourcekitd: self.sourcekitd)
+    skreq[keys.request] = requests.diagnostics
+    skreq[keys.sourcefile] = snapshot.uri.pseudoPath
+
+    // FIXME: SourceKit should probably cache this for us.
+    skreq[keys.compilerargs] = compilerArgs
+
+    let dict = try await self.sourcekitd.send(skreq, fileContents: snapshot.text)
+
+    try Task.checkCancellation()
+    guard (try? documentManager.latestSnapshot(snapshot.uri).id) == snapshot.id else {
+      // Check that the document wasn't modified while we were getting diagnostics. This could happen because we are
+      // calling `fullDocumentDiagnosticReport` from `publishDiagnosticsIfNeeded` outside of `messageHandlingQueue`
+      // and thus a concurrent edit is possible while we are waiting for the sourcekitd request to return a result.
+      throw ResponseError.unknown("Document was modified while loading document")
+    }
+
+    let supportsCodeDescription = self.clientHasDiagnosticsCodeDescriptionSupport
+    var diagnostics: [Diagnostic] = []
+    dict[keys.diagnostics]?.forEach { _, diag in
+      if let diag = Diagnostic(diag, in: snapshot, useEducationalNoteAsCode: supportsCodeDescription) {
+        diagnostics.append(diag)
+      }
+      return true
+    }
+
+    return RelatedFullDocumentDiagnosticReport(items: diagnostics)
+  }
+
+  func requestFallbackReport(with snapshot: DocumentSnapshot) async throws
```

Comment:
i put all the private funcs in a `private extension`, but there is still a compile error. `method must be declared private because its result uses a private type`. let me move all the funcs back to the actor and mark them private.